### ISSUE TITLE
Summarize contacts-v1 optimization experiments

### ISSUE THREAD

**Issue by eric-czech** [Open-Athena/MarinFold#154]
Issue to summarize results across several recent, related contacts-v1 experiments.

## Experiments

| Experiment | Checkpoint pattern | Hugging Face | W&B[^wandb] |
|:--|:--|:--:|:--:|
| [#75](https://github.com/Open-Athena/MarinFold/issues/75) · [code](https://github.com/marin-community/marin/blob/eac/plm-exp75/experiments/protein/exp75_sweep.py)<br><sub>1.5B pilot</sub> | `gcs://us-east5/checkpoints/protein/prot-exp75-*` | [exp75](https://huggingface.co/open-athena/marinfold-exp75) | [runs](https://wandb.ai/eric-czech/marin/groups/exp75-contacts-v1-tune) |
| [#117](https://github.com/Open-Athena/MarinFold/issues/117) · [code](https://github.com/marin-community/marin/blob/eac/plm-exp117/experiments/protein/exp117_sweep.py)<br><sub>1.5B tuning</sub> | <code>gcs://(us-east1&#124;us-east5&#124;us-west4&#124;europe-west4)/checkpoints/protein/prot-exp117-*</code> | [exp117](https://huggingface.co/open-athena/marinfold-exp117) | [runs](https://wandb.ai/eric-czech/marin/groups/exp117-contacts-v1-tune) |
| [#146](https://github.com/Open-Athena/MarinFold/issues/146) · [code](https://github.com/marin-community/marin/blob/eac/plm-exp146/experiments/protein/exp146_sweep.py)<br><sub>3B tuning</sub> | <code>gcs://(us-east1&#124;us-east5&#124;us-west4&#124;europe-west4)/checkpoints/protein/prot-exp146-*</code> | [exp146](https://huggingface.co/open-athena/marinfold-exp146) | [runs](https://wandb.ai/eric-czech/marin/groups/exp146-contacts-v1-3b-tune) |
| [#153](https://github.com/Open-Athena/MarinFold/issues/153) · [code](https://github.com/marin-community/marin/blob/eac/plm-exp153-cw/experiments/protein/exp153_sweep.py)<br><sub>6B tuning</sub> | `s3://marin-us-east-02a/MarinFold/exp154_qwen_contacts_v1/checkpoints/prot-exp153-*` | — | [runs](https://wandb.ai/eric-czech/marin/groups/exp153-contacts-v1-6b-tune) |
| [#166](https://github.com/Open-Athena/MarinFold/issues/166) · [code](https://github.com/marin-community/marin/blob/eac/plm-exp166/experiments/protein/exp166_sweep.py)<br><sub>AA augmentation</sub> | <code>gcs://(us-east1&#124;us-east5&#124;europe-west4)/checkpoints/protein/prot-exp166-*</code> | [exp166](https://huggingface.co/open-athena/marinfold-exp166) | [runs](https://wandb.ai/eric-czech/marin/groups/exp166-contacts-v1-aa-augmentation) |

[^wandb]: W&B links currently point to `eric-czech/marin`. Public links under the `open-athena` team are blocked until model seats are available.

## Results

1. [Tuning Overview](https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5192141278) — scaling results and hyperparameter dependence
2. [Training Stability](https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5194394007) — learning-rate risk/reward and weight-decay interaction
3. [Sequence Augmentation](https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5197672867) — de novo and continued-pretraining comparisons
4. [R-precision Trajectory](https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5227125015) — 3B vs 1.5B evals over the course of training

**Comment by eric-czech** [comment]
## Tuning Overview

Collected 129 finished W&B runs ([csv](https://github.com/Open-Athena/MarinFold/blob/560d530296236bdb9bf9a2ff2eabf976bfa7adc1/experiments/exp154_models_contacts_v1_parameter_scaling/data/wandb_runs.csv)) from the latest sweep subversions of #75, #117, and #146. The 1.5B sweep is the only group with enough coverage to model across epochs. The 1.5B eight-epoch results provide the strongest direct evidence. The other groups are included to report their best measured configurations. The analysis code and data are in [`560d530`](https://github.com/Open-Athena/MarinFold/commit/560d530296236bdb9bf9a2ff2eabf976bfa7adc1).

### Validation loss by training epochs

The plot below includes every finished run. Rings mark the lowest validation loss for each model size and epoch count.

<p align="center">
<img width="768px" src="https://github.com/Open-Athena/MarinFold/blob/560d530296236bdb9bf9a2ff2eabf976bfa7adc1/experiments/exp154_models_contacts_v1_parameter_scaling/plots/validation_loss_vs_epochs.png?raw=true"/>
</p>

*Data*: [validation_loss_vs_epochs.csv](https://github.com/Open-Athena/MarinFold/blob/560d530296236bdb9bf9a2ff2eabf976bfa7adc1/experiments/exp154_models_contacts_v1_parameter_scaling/data/validation_loss_vs_epochs.csv)

### Divergent runs

Five 1.5B eight-epoch configurations have unusually high final training and validation loss. They are excluded from the GP fit.

![Divergence screen](https://github.com/Open-Athena/MarinFold/blob/560d530296236bdb9bf9a2ff2eabf976bfa7adc1/experiments/exp154_models_contacts_v1_parameter_scaling/plots/divergence_screen.png?raw=true)

### 1.5B hyperparameter dependence

A Matérn GP was fit to 105 unique, non-divergent 1.5B configurations. The partial-dependence curves show local effects within the tested ranges. They provide rough guidance rather than an extrapolating model.

![GP partial dependence](https://github.com/Open-Athena/MarinFold/blob/560d530296236bdb9bf9a2ff2eabf976bfa7adc1/experiments/exp154_models_contacts_v1_parameter_scaling/plots/gpr_partial_dependence.png?raw=true)

### Best measured configurations

Each row here gives the lowest observed validation loss for that model size and epoch count.

| Model | Epochs | Learning rate | Weight decay | Batch size |
|---|---:|---:|---:|---:|
| 1.5B | 1 | `7e-4` | `0.05` | 128 |
| 1.5B | 2 | `7e-4` | `0.8` | 128 |
| 1.5B | 4 | `1e-3` | `0.05` | 128 |
| 1.5B | 8 | `3.1623e-3` | `0.2` | 64 |
| 1.5B | 16 | `3.1623e-3` | `0.2` | 256 |
| 3B | 2 | `3.1623e-3` | `0.2` | 64 |
| 3B | 8 | `3.1623e-3` | `0.4` | 256 |

**Comment by timodonnell** [comment]
Thanks @eric-czech . These are all cosine decay LR, right? Did you experiment with WSD? I am wondering if we can do something like WSD (or [WSD-S](https://arxiv.org/abs/2410.05192)) for a larger run on our ESMFold2 data where we don't know the total epochs we will have time for up front.

**Comment by eric-czech** [comment]
They're all cosine, yea.  I haven't tried WSD on this dataset (just the old protein docs), but it would make more sense to me to prefer that on a bigger dataset.

**Comment by eric-czech** [comment]
## Training Stability

This analysis covers 42 finished runs from the latest [exp117](https://github.com/Open-Athena/MarinFold/issues/117) sweep subversion. All runs use the 1.5B model and eight training epochs. The analysis code and data are in [`33230dc`](https://github.com/Open-Athena/MarinFold/commit/33230dc8c9c7130771fe6edc1698dd21c55a650b).

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/33230dc8c9c7130771fe6edc1698dd21c55a650b/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp117_lr_training_validation.png?raw=true" alt="Training and validation losses by learning rate">
</p>

*Data*: [exp117_lr_training_validation.csv](https://github.com/Open-Athena/MarinFold/blob/33230dc8c9c7130771fe6edc1698dd21c55a650b/experiments/exp154_models_contacts_v1_parameter_scaling/data/exp117_lr_training_validation.csv)

The learning-rate sweep shows a clear risk/reward tradeoff. Increasing LR improves the best observed validation loss and broadens the distribution of outcomes through `3.16e-3`. At `1e-2`, typical performance degrades without improving the best result. The boxes and diamonds summarize Tukey inliers at each LR; triangles show the excluded outliers.

The LR to emphasize depends on the available run budget:

| Planned runs | LR | WD | Rationale |
|---:|---:|---:|---|
| 1 | `3.16e-4` | `1.6` | Best choice when only attempting a single run |
| About 10 | `1e-3` | `0.8` | Best choice if you're going to at least tune WD a little and/or run multiple seeds |
| 30 or more | `3.16e-3` | `0.2` | Best choice for maximum performance w/ more tuning and/or trial replicates |

### Learning rate x weight decay

<p align="center">
  <img width="640px" src="https://github.com/Open-Athena/MarinFold/blob/33230dc8c9c7130771fe6edc1698dd21c55a650b/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp117_lr_weight_decay_heatmap.png?raw=true" alt="Best validation loss by learning rate and weight decay">
</p>

*Data*: [exp117_lr_weight_decay_heatmap.csv](https://github.com/Open-Athena/MarinFold/blob/33230dc8c9c7130771fe6edc1698dd21c55a650b/experiments/exp154_models_contacts_v1_parameter_scaling/data/exp117_lr_weight_decay_heatmap.csv)

This was an adaptive search rather than a complete factorial sweep. Convergence required finding a configuration with no known better loss that was also non-inferior to every neighboring point on the three-dimensional LR × weight-decay × batch-size grid. Blank cells are combinations that the search did not evaluate.

The best weight decay decreases as LR increases: `1.6` at LR `3.16e-4`, `0.8` at `1e-3`, `0.2` at `3.16e-3`, and `0.1` at `1e-2`. Higher learning rates therefore worked best with less weight decay in this sweep.

The best observed configuration was LR `3.16e-3`, weight decay `0.2`, and batch size `64`, with validation loss `2.713`. It is an attractive target when enough runs are available to tolerate the wider outcome distribution at this LR.

**Comment by eric-czech** [comment]
## Sequence Augmentation

Experiment [#166](https://github.com/Open-Athena/MarinFold/issues/166) tests a training-only augmentation of contacts-v1 documents. Each position–amino-acid statement remains intact, but the statements within the sequence section are reshuffled every time a training document is encountered. Position assignments, contact targets, document boundaries, and validation data do not change.

The analysis code and data are available in [`2ba6d00`](https://github.com/Open-Athena/MarinFold/commit/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778).

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp166_augmentation_schematic.png?raw=true" alt="How sequence augmentation changes contacts-v1 documents">
</p>

### Validation-loss distributions

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp166_epoch_distributions.png?raw=true" alt="Validation-loss distributions for exp117 and exp166">
</p>

*Data*: [exp166_epoch_distributions.csv](https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/data/exp166_epoch_distributions.csv)

The best warm-started augmented run reached `2.664`, compared with `2.704` for the best exp117 16-epoch run and `2.713` for the best exp117 8-epoch run. These groups contain different hyperparameter configurations, so the distributions provide context rather than a controlled comparison.

### Matched configurations

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp166_matched_final_loss.png?raw=true" alt="Matched exp117 and exp166 validation losses">
</p>

*Data*: [exp166_final_comparisons.csv](https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/data/exp166_final_comparisons.csv)

This comparison holds learning rate, weight decay, and batch size fixed. Each arrow ends at the best finished exp166 attempt for that configuration; open diamonds show other finished regional attempts.

Continued pretraining with augmentation improved four of six configurations, with a median validation-loss change of `-0.040`. Training from scratch improved only two of six, with a median change of `+0.012`. The scratch result is not conclusive because the hyperparameters were selected on the original training distribution. Establishing whether augmentation helps from scratch requires retuning for the augmented distribution.  By the same logic, the continued-pretraining results probably could have been improved.

Several continued-pretraining runs were unstable after the augmentation switch. Training loss rose gradually before decreasing again, while validation loss sometimes spiked before recovering. This behavior is consistent with an abrupt distribution shift. A follow-up should blend original and augmented examples during the transition, or gradually increase the augmentation probability.

<details>
<summary>Validation trajectories</summary>

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/2ba6d00b24a2c9ac0ad7af61a4b1578659c5e778/experiments/exp154_models_contacts_v1_parameter_scaling/plots/exp166_validation_trajectories.png?raw=true" alt="Validation trajectories for matched exp117 and exp166 configurations">
</p>

</details>

Overall, continued pretraining with augmentation improved the tested 1.5B configurations. The from-scratch results remain uncertain until the sweep is retuned for the augmented training distribution.

**Comment by eric-czech** [comment]
## R-precision Trajectory

This analysis follows 24 checkpoints from 3 contacts-v1 runs, primarily to investigate why 3B models have better loss but worse R-precision than the best 1.5B models. The [1.5B E8 final checkpoint](https://huggingface.co/open-athena/marinfold-exp117/tree/main/prot-exp117-cv1-s02-1_5b-e8-lr3p162e-3-wd0p2-bs64-europe-west4/hf/step-71359) from [exp117](https://github.com/Open-Athena/MarinFold/issues/117) serves as the control. The comparison configurations add either training epochs with [1.5B E16](https://huggingface.co/open-athena/marinfold-exp117/tree/main/prot-exp117-cv1-s02-1_5b-e16-lr3p162e-3-wd0p2-bs256-europe-west4/hf/step-35679), also from [exp117](https://github.com/Open-Athena/MarinFold/issues/117), or parameters with [3B E8](https://huggingface.co/open-athena/marinfold-exp146/tree/main/prot-exp146-cv1-s01-3b-e8-lr3p162e-3-wd0p4-bs256-us-east1/hf/step-17839) from [exp146](https://github.com/Open-Athena/MarinFold/issues/146).

The analysis code and data are available in [`d9a4939`](https://github.com/Open-Athena/MarinFold/commit/d9a4939ee9762d69eddad6810d1ed4ea5c25d197).

<p align="center">
  <img src="https://github.com/Open-Athena/MarinFold/blob/d9a4939ee9762d69eddad6810d1ed4ea5c25d197/experiments/exp169_evals_selected_checkpoints_117_146/plots/checkpoint_trajectory.png?raw=true" alt="Validation loss and R-precision over training">
</p>

*Data*: [trajectory_checkpoint_metrics.csv](https://github.com/Open-Athena/MarinFold/blob/d9a4939ee9762d69eddad6810d1ed4ea5c25d197/experiments/exp169_evals_selected_checkpoints_117_146/data/trajectory_checkpoint_metrics.csv)

At approximately 37.4B training tokens, the 3B run reaches `0.5077` all-range R-precision, ahead of the 1.5B E8 control at `0.4944` and the still-in-progress E16 run at `0.4233`. Validation loss and R-precision both improve through the final 3B checkpoint, providing no evidence of contact overfitting.  The E16 run continues to 74.8B tokens and reaches `0.5338`.

**Conclusion**: The 3B model doesn't overfit, so it's not clear why scaling models beyond 1.5B is unproductive.
